### PR TITLE
fix checking git submodules during a commit hook

### LINKED ISSUE #125954: With submodules=true, using `./x.py fmt` as a git hook makes bootstrap delete all submodules
To reproduce, I am in a worktree without `.git` (i.e., not the "main" worktree). I have a pre-commit hook set up as follows
```
#!/bin/sh
./x.py fmt || true # ignore failures
```
Now I change something trivial (e.g. remove a trailing `.` in some comments) and commit. That does the following
```
$ git commit -am test
Building bootstrap
    Finished `dev` profile [unoptimized] target(s) in 0.06s
Updating submodule src/tools/cargo
Saved working directory and index state WIP on from-ref: 64e7337be59 more explicitly state the basic rules of working with the obtained raw pointers
```
After a second or two I get suspicious and hit Ctrl-C. Now it looks like bootstrap has done something with the (previously clean) cargo submodule
```
$ git status
On branch from-ref
Your branch is up to date with 'ralf/from-ref'.

Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
  (commit or discard the untracked or modified content in submodules)
    modified:   library/core/src/ptr/mod.rs
    modified:   src/tools/cargo (modified content, untracked content)
```
So let's `cd src/tools/cargo` and see what's up
```
$ git status
HEAD detached at 7a6fad098
Changes not staged for commit:
  (use "git add <file>..." to update what will be committed)
  (use "git restore <file>..." to discard changes in working directory)
    modified:   .github/ISSUE_TEMPLATE/config.yml
    modified:   .gitignore
    modified:   CONTRIBUTING.md
    modified:   Cargo.lock
    modified:   Cargo.toml
    modified:   LICENSE-APACHE
    modified:   README.md
    modified:   triagebot.toml

Untracked files:
  (use "git add <file>..." to include in what will be committed)
    .editorconfig
    .git-blame-ignore-revs
    .gitattributes
    .github/ISSUE_TEMPLATE/blank_issue.md
    .github/ISSUE_TEMPLATE/bug_report.md
    .github/ISSUE_TEMPLATE/diagnostics.yaml
    .github/ISSUE_TEMPLATE/documentation.yaml
    .github/ISSUE_TEMPLATE/ice.md
    .github/ISSUE_TEMPLATE/ice.yaml
    .github/ISSUE_TEMPLATE/library_tracking_issue.md
    .github/ISSUE_TEMPLATE/regression.md
    .github/ISSUE_TEMPLATE/tracking_issue.md
    .github/pull_request_template.md
    .github/workflows/ci.yml
    .github/workflows/dependencies.yml
    .gitmodules
    .mailmap
    .reuse/
    COPYRIGHT
    INSTALL.md
    LICENSES/
    RELEASES.md
    compiler/
    config.example.toml
    configure
    library/
    rust-bors.toml
    rustfmt.toml
    src/README.md
    src/bootstrap/
    src/ci/
    src/doc/complement-design-faq.md
    src/doc/complement-lang-faq.md
    src/doc/complement-project-faq.md
    src/doc/favicon.inc
    src/doc/footer.inc
    src/doc/full-toc.inc
    src/doc/grammar.md
    src/doc/guide-crates.md
    src/doc/guide-error-handling.md
    src/doc/guide-ffi.md
    src/doc/guide-macros.md
    src/doc/guide-ownership.md
    src/doc/guide-plugins.md
    src/doc/guide-pointers.md
    src/doc/guide-strings.md
    src/doc/guide-tasks.md
    src/doc/guide-testing.md
    src/doc/guide-unsafe.md
    src/doc/guide.md
    src/doc/index.md
    src/doc/intro.md
    src/doc/man/rustc.1
    src/doc/man/rustdoc.1
    src/doc/not_found.md
    src/doc/redirect.inc
    src/doc/reference.md
    src/doc/robots.txt
    src/doc/rust.css
    src/doc/rust.md
    src/doc/rustc/
    src/doc/rustdoc.md
    src/doc/rustdoc/
    src/doc/style-guide/
    src/doc/tutorial.md
    src/doc/unstable-book/
    src/doc/version_info.html.template
    src/etc/CONFIGS.md
    src/etc/cat-and-grep.sh
    src/etc/completions/
    src/etc/cpu-usage-over-time-plot.sh
    src/etc/ctags.rust
    src/etc/dec2flt_table.py
    src/etc/gdb_load_rust_pretty_printers.py
    src/etc/gdb_lookup.py
    src/etc/gdb_providers.py
    src/etc/generate-deriving-span-tests.py
    src/etc/generate-keyword-tests.py
    src/etc/htmldocck.py
    src/etc/indenter
    src/etc/installer/
    src/etc/lldb_batchmode.py
    src/etc/lldb_commands
    src/etc/lldb_lookup.py
    src/etc/lldb_providers.py
    src/etc/natvis/
    src/etc/pre-push.sh
    src/etc/rust-gdb
    src/etc/rust-gdbgui
    src/etc/rust-lldb
    src/etc/rust-windbg.cmd
    src/etc/rust_analyzer_settings.json
    src/etc/rust_types.py
    src/etc/test-float-parse/
    src/etc/third-party/
    src/librustdoc/
    src/rustdoc-json-types/
    src/stage0
    src/tools/
    src/version
    tests/COMPILER_TESTS.md
    tests/assembly/
    tests/auxiliary/
    tests/codegen-units/
    tests/codegen/
    tests/coverage-run-rustdoc/
    tests/coverage/
    tests/crashes/
    tests/debuginfo/
    tests/incremental/
    tests/mir-opt/
    tests/pretty/
    tests/run-make-fulldeps/
    tests/run-make/
    tests/run-pass-valgrind/
    tests/rustdoc-gui/
    tests/rustdoc-js-std/
    tests/rustdoc-js/
    tests/rustdoc-json/
    tests/rustdoc-ui/
    tests/rustdoc/
    tests/ui-fulldeps/
    tests/ui/
    x
    x.ps1
    x.py

no changes added to commit (use "git add" and/or "git commit -a")
```

Eh... so the added files are rustc files? Why are those being added in `src/tools/cargo`?
Something is going seriously wrong here.

### PR DESCRIPTION
**Pull request by RalfJung**
Fixes https://github.com/rust-lang/rust/issues/125954
The one thing that still puzzles me is why this only is a problem with worktrees.

r? @onur

### CODE REVIEW

**Comment by RalfJung** [comment]
Hm... Any idea why that command would fail...?

**Comment by onur-ozkan** [comment]
To reproduce the problem, you can enter the container in interactive mode with `./src/ci/docker/run.sh x86_64-gnu-llvm-17 --dev`.

Running the failed command from https://github.com/rust-lang/rust/pull/126255#issuecomment-2159997223 gives the following error message:

```sh
nimda@b536d58a7363:~$ git --git-dir /checkout/.git rev-list --author=bors@rust-lang.org -n1 --first-parent HEAD -- /checkout/src/llvm-project /checkout/src/bootstrap/download-ci-llvm-stamp /checkout/src/version
fatal: /checkout/src/llvm-project: '/checkout/src/llvm-project' is outside repository at '/checkout/obj'
```

As far as I understand you cannot access another repository (or submodule) while explicitly setting `--git-dir`.

**Comment by RalfJung** [comment]
That can't be quite right, `git --git-dir .git rev-list -n1 --first-parent HEAD -- src/tools/cargo` works just fine here.

**Comment by onur-ozkan** [comment]
> To reproduce the problem, you can enter the container in interactive mode with `./src/ci/docker/run.sh x86_64-gnu-llvm-17 --dev`.
> 
> Running the failed command from [#126255 (comment)](https://github.com/rust-lang/rust/pull/126255#issuecomment-2159997223) gives the following error message:
> 
> ```shell
> nimda@b536d58a7363:~$ git --git-dir /checkout/.git rev-list --author=bors@rust-lang.org -n1 --first-parent HEAD -- /checkout/src/llvm-project /checkout/src/bootstrap/download-ci-llvm-stamp /checkout/src/version
> fatal: /checkout/src/llvm-project: '/checkout/src/llvm-project' is outside repository at '/checkout/obj'
> ```
> 
> As far as I understand you cannot access another repository (or submodule) while explicitly setting `--git-dir`.

fwiw changing current directory to `/checkout` from `/checkout/obj` fixed the problem on my side.

**Comment by RalfJung** [comment]
I don't understand why changing the current directory makes a difference when `--git-dir` is set... but that's exactly what my most recent commit also does.

**Comment by RalfJung** [comment]
I am definitely not understanding `--git-dir` properly. I played around with this
```
git --git-dir ~/src/rust/rustc/.git rev-list -n1 --first-parent HEAD -- ~/src/rust/rustc/src/tools/cargo
```
and depending on my working directory, this sometimes shows a commit and sometimes not. And when my working dir is a different git checkout it just fails.

So do we have to do both, change the working dir and set `--git-dir`?!?

**Comment by RalfJung** [comment]
Ah, [this](https://stackoverflow.com/questions/7149984/how-do-i-execute-a-git-command-without-being-in-the-repository) explains it, partially. So we definitely need the working dir, or set `-C`.

But then also setting `--git-dir` seems like a hack. Maybe I should figure out which env var is the problem and just unset that.

**Comment by onur-ozkan** [comment]
> Maybe I should figure out which env var is the problem and just unset that.

This could be the best solution. Adding `--git-dir` individually on multiple places makes things more complicated than they should be.

**Comment by RalfJung** [comment]
I found something that works, but it still behaves a bit strange -- when I actually have an intended submodule change and `git add && git commit` that, then the `x.py` invocation in the commit hook will first check out the old cargo and then, when it runs a second time, it checks out the commit I set it to...

Maybe we should just skip all submodule processing when GIT_DIR is set? That happens during a git invocation and doing advanced git surgery in the middle of that might just not be a good idea.

**Comment by onur-ozkan** [comment]
Would setting `--git-dir` next to every `current_dir` usage solve this problem? If this works, it would be much easier to maintain as it only adds one extra argument to the git command without significantly increasing complexity.

**Comment by onur-ozkan** [comment]
> Would setting `--git-dir` next to every `current_dir` usage solve this problem? If this works, it would be much easier to maintain as it only adds one extra argument to the git command without significantly increasing complexity.

I think you already did that before the latest change, right? The main issue was that we are manually constructing git commands all over the bootstrap codebase. If we manage all the git commands from a single interface, we could maintain this much more easily.

**Comment by RalfJung** [comment]
I tried that, yeah -- but it's still not great in a commit hook: when you want to actually change the submodule, and then commit, weird stuff happens. It first un-does your change and then un-does the un-doing, at least with the cargo subrepo which is currently checked twice each time (first by the loop that always initializes cargo, then by the "check all existing submodules" check).

**Comment by RalfJung** [comment]
Also to be clear I have not systematically searched for git invocations in the codebase, I just refactored the one I stumbled into.

**Comment by onur-ozkan** [comment]
Can we wait until we ship https://github.com/rust-lang/rust/pull/126309 ? It should help significantly in making this work cleaner.

**Comment by RalfJung** [comment]
Sure, this is not urgent.

**Comment by onur-ozkan** [comment]
> Can we wait until we ship #126309 ? It should help significantly in making this work cleaner.

It is merged.

**Comment by onur-ozkan** [comment]
We had (this)[https://github.com/rust-lang/rust/pull/77683] workaround previously, so if we unset GIT_DIR in bootstrap (in the centrealized git function in helper module), that should fix the problem. I am not sure why that undoing->un-undoing thing happens, but it doesn't break anything in the process as far as I understand..

**Comment by RalfJung** [comment]
Un-setting GIT_DIR is not enough, I tried that. One has to also unset GIT_INDEX_FILE.

>  I am not sure why that undoing->un-undoing thing happens, but it doesn't break anything in the process as far as I understand..

I think it would break something for submodules that are only updated once, i.e. all the ones which are not forcefully checked out. (Ideally bootstrap would be patched to not update anything twice...)

File: src/bootstrap/src/lib.rs [hunk 1]
```diff
@@ -520,19 +520,20 @@ impl Build {
             return;
         }
 
-        // check_submodule
-        let checked_out_hash =
-            output(helpers::git(Some(&absolute_path)).args(["rev-parse", "HEAD"]));
-        // update_submodules
+        let submodule_git = || helpers::git(Some(&absolute_path));
```
Comment (RalfJung) [inline src/bootstrap/src/lib.rs:523]: I left this closure in since I think it makes the code more readable to abstract this common call away.

**Review by onur-ozkan — APPROVED** [review]
Thanks for the fix!

> Ideally bootstrap would be patched to not update anything twice...

We are planning to track and cache git certain invocations, I think by using the same interface we can achieve this.

**Comment by RalfJung** [comment]
> We are planning to track and cache git certain invocations, I think by using the same interface we can achieve this.

For this particular case, it seems easier to make `update_existing_submodules` return a list of submodules (instead of calling `update_submodule` directly), add `["src/tools/cargo", "src/doc/book", "library/backtrace", "library/stdarch"]` if they aren't in there already, and then iterate the result. :shrug:

**Pull request merged**